### ISSUE TITLE
Rethink log-Mel as the timing-sensitive audio frontend

### ISSUE THREAD

**Issue by s3d-i** [Pulsefield/Pulsefield-model#13]
## Problem

Pulsefield currently commits to a fairly aggressive audio bottleneck before the model sees the signal. The default frontend in [`mel_base.py`](https://github.com/Pulsefield/Pulsefield-model/blob/main/src/pulsefield_model/features/mel_base.py) is:

- 16 kHz mono audio;
- 400-sample / 25 ms Hann STFT window;
- 160-sample / 10 ms hop (100 Hz);
- `power=2.0`, so complex STFT phase is discarded;
- 80 fixed Mel bands over 20 Hz–8 kHz;
- log compression after flooring.

This representation is compact and convenient, but it was adopted without establishing that it preserves the information required by a timing-sensitive rhythm-game mapper. The concern is not simply that "100 Hz means timing is quantized to 10 ms"—overlapping windows can encode sub-hop offsets. The more serious question is whether the combination of **phase removal, Mel frequency pooling, 25 ms temporal smearing, downsampling, and log compression** creates an information ceiling for onset localization / microtiming that the downstream model cannot recover from.

For Pulsefield, this matters because the relevant target may be much finer than ordinary beat/downbeat tracking: exact note/onset alignment, dense high-BPM passages, relative microtiming, and potentially millisecond-scale distinctions that affect chart feel.

## Important correction: BeatThis does use compressed log-Mel

BeatThis is **not** evidence that strong beat trackers avoid Mel preprocessing. Its official [`LogMelSpect`](https://github.com/CPJKU/beat_this/blob/main/beat_this/preprocessing.py) uses:

- 22.05 kHz audio;
- `n_fft=1024`;
- `hop_length=441` = 20 ms / 50 Hz;
- 128 Slaney Mel bins;
- magnitude (`power=1`) followed by `log1p(1000 * x)`.

Its training preprocessing script explicitly computes and stores these log-Mel spectrograms before training: [`launch_scripts/preprocess_audio.py`](https://github.com/CPJKU/beat_this/blob/main/launch_scripts/preprocess_audio.py).

So the hypothesis should be stated more narrowly:

> BeatThis demonstrates that a 128-bin, 50 Hz log-Mel representation can be sufficient for strong **conventional beat/downbeat tracking**. It does not establish that Mel is a sufficient statistic for Pulsefield's stricter **sub-frame onset / phase / microtiming** objective.

This is a reason to measure the frontend ceiling, not a reason to assume Mel is bad.

### 1. Phase removal

The current path performs roughly

```text
complex STFT -> |X|^2 -> Mel projection -> log
```

Many different waveforms can have the same or nearly the same magnitude spectrum while differing substantially in relative phase, transient shape, and within-window temporal concentration. Those distinctions may matter for localizing attacks precisely.

Raw phase angle is probably not the right model feature because it wraps and is highly translation-sensitive. More useful candidates include temporal phase differences / instantaneous-frequency-like features and frequency-direction phase gradients / group-delay-like timing features.

### 2. Mel pooling

The Mel filterbank is many-to-one across linear-frequency bins. Fine harmonic, broadband, and transient structures can collapse to the same band energy. This may be harmless for semantic audio recognition while still deleting cues useful for onset timing.

### 3. Window smearing vs hop rate

The current 10 ms hop and 25 ms window are two separate constraints:

- hop controls temporal sampling density;
- window length controls how much a sharp transient is integrated / spread in time.

A 100 Hz frame rate is therefore not by itself a hard ±5 ms localization bound, but increasing hop rate alone may also fail if the dominant loss comes from the analysis window or phase destruction.

### 4. Downsampling / high-frequency transient cues

16 kHz audio removes content above 8 kHz. Whether that matters for mapping is unknown, but high-frequency attacks are a plausible timing cue and should be tested rather than assumed irrelevant.

## Candidate representations to ablate

Do not jump directly to a large learned frontend. Start with an interpretable information-loss ladder:

1. **complex linear STFT** (`Re/Im`) — high-information reference;
2. **linear-frequency magnitude STFT** — isolates the cost of phase removal;
3. **Mel magnitude/log-Mel** — isolates the additional cost of Mel pooling;
4. **magnitude + phase-gradient features** — retain timing-relevant phase structure without raw phase angles;
5. **multi-resolution STFT** — short/high-rate transient branch plus longer-window spectral/context branch.

Later candidates, only if the simple ablations justify them:

- time-frequency reassignment / reassigned spectrogram features;
- learned Gabor/LEAF-like frontend;
- multi-rate frontend where a small timing branch runs at 200–1000 Hz while spectral context stays much slower.

The key idea is that **high temporal rate does not require carrying the full spectral representation at the same rate**.

### A. Timing information-loss ladder

Build a deliberately small probe whose only job is to recover precise event time from each representation. Keep model capacity and training data fixed while changing only the frontend.

Start with controlled signals and then move toward real music:

- clicks / impulses with continuously sampled sub-frame offsets;
- kick, snare, hi-hat and other transient samples at controlled offsets;
- polyphonic mixtures and music excerpts with injected / known transient positions;
- eventually real annotated mapper-relevant events.

Measure continuous timing error rather than only frame classification:

```text
MAE(t)
P(|error| < 1 ms)
P(|error| < 2.5 ms)
P(|error| < 5 ms)
P(|error| < 10 ms)
```

Vary **window length and hop independently**. A useful matrix is, for example, 50 / 100 / 200 / 500 Hz hops crossed with several analysis-window lengths. This prevents us from incorrectly attributing window blur to frame rate.

### B. Mel metamer / null-space experiment

Construct pairs of waveforms that are indistinguishable or nearly indistinguishable under the exact Pulsefield Mel frontend but differ strongly in a timing-relevant property.

Given reference audio `x`, optimize another waveform `x'` such that:

```text
Mel(x') ~= Mel(x)
```

while maximizing a timing-relevant difference, for example:

```text
|transient_centroid(x') - transient_centroid(x)|
```

or a phase/transient-structure distance.

If two signals can have near-identical `80-bin @ 100 Hz` Pulsefield features while humans or a higher-information onset detector reliably distinguish their attack timing, then the frontend has a directly demonstrated task-relevant null space.

Repeat this at 50 / 100 / 200 / 500 Hz to estimate how the maximum recoverably-hidden timing displacement changes with temporal resolution.

### C. Downstream mapper ablation

Only after the probe establishes a difference, test the most promising frontends in Mapper under matched model capacity / compute budget. Compare not just token CE but rollout-relevant timing metrics:

- onset alignment error;
- dense/high-BPM sections;
- longest local timing deviation;
- beat-relative phase error;
- generated chart quality under otherwise identical decoding.

## Decision criteria

Do **not** replace Mel because it is theoretically lossy—all useful representations are lossy. Replace or augment it only if the discarded information is demonstrably useful for the task.

- If log-Mel matches complex/linear representations on fine-timing probes and downstream mapping, keep it: the compression is valuable.
- If linear STFT wins but complex STFT does not add more, the main problem is likely Mel pooling rather than phase.
- If phase-aware features materially reduce sub-frame error, retain a compact phase/timing branch.
- If shorter windows help substantially, use a multi-resolution frontend rather than simply increasing the global frame rate.
- Any gain should be reported against memory, bandwidth, encoder FLOPs, and attention cost.

## Acceptance criteria

- [ ] Quantify the fine-timing ceiling of the current 16 kHz / 80-bin / 100 Hz / 25 ms log-Mel frontend.
- [ ] Separate the effects of hop rate, window length, phase removal, and Mel pooling.
- [ ] Add at least one higher-information reference (linear magnitude or complex STFT).
- [ ] Produce a Mel-metamer/null-space example using the exact current frontend.
- [ ] Report continuous sub-frame timing metrics, not only frame-level accuracy.
- [ ] Decide whether to keep, augment, or replace Mel based on measured task performance and compute cost.

## Non-goal

This issue is **not** claiming that Mel is generally unsuitable for rhythm or beat tracking. BeatThis is a direct counterexample to that broad claim. The question is specifically whether the current Mel bottleneck is appropriate for Pulsefield's finer timing requirements.
